server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1450,14 +1452,57 @@ private void cleanupOldShardGens(RepositoryData existingRepositoryData, Reposito
     }
 
     @Override
-    public SnapshotInfo getSnapshotInfo(final SnapshotId snapshotId) {
-        try {
-            return SNAPSHOT_FORMAT.read(blobContainer(), snapshotId.getUUID(), namedXContentRegistry);
-        } catch (NoSuchFileException ex) {
-            throw new SnapshotMissingException(metadata.name(), snapshotId, ex);
-        } catch (IOException | NotXContentException ex) {
-            throw new SnapshotException(metadata.name(), snapshotId, "failed to get snapshots", ex);
+    public void getSnapshotInfo(GetSnapshotInfoContext context) {
+        // put snapshot info downloads into a task queue instead of pushing them all into the queue to not completely monopolize the
+        // snapshot meta pool for a single request
+        final int workers = Math.min(threadPool.info(ThreadPool.Names.SNAPSHOT_META).getMax(), context.snapshotIds().size());
+        final BlockingQueue<SnapshotId> queue = new LinkedBlockingQueue<>(context.snapshotIds());
+        for (int i = 0; i < workers; i++) {
+            getOneSnapshotInfo(queue, context);
+        }
+    }
+
+    /**
+     * Tries to poll a {@link SnapshotId} to load {@link SnapshotInfo} for from the given {@code queue}.
+     */
+    private void getOneSnapshotInfo(BlockingQueue<SnapshotId> queue, GetSnapshotInfoContext context) {
+        final SnapshotId snapshotId = queue.poll();
+        if (snapshotId == null) {
+            return;
         }
+        threadPool.executor(ThreadPool.Names.SNAPSHOT_META).execute(() -> {
+            if (context.done()) {
+                return;
+            }
+            if (context.isCancelled()) {
+                queue.clear();
+                context.onFailure(new TaskCancelledException("task cancelled"));
+                return;
+            }
+            Exception failure = null;
+            SnapshotInfo snapshotInfo = null;
+            try {
+                snapshotInfo = SNAPSHOT_FORMAT.read(blobContainer(), snapshotId.getUUID(), namedXContentRegistry);
+            } catch (NoSuchFileException ex) {
+                failure = new SnapshotMissingException(metadata.name(), snapshotId, ex);
+            } catch (IOException | NotXContentException ex) {
+                failure = new SnapshotException(metadata.name(), snapshotId, "failed to get snapshots", ex);
```

Comment:
```suggestion
                failure = new SnapshotException(metadata.name(), snapshotId, "failed to get snapshot info for " + snapshotId, ex);
```

Reply:
No need to add the snapshot id to the message IMO, we already have it as a field on the snapshot exception and that will be serialized. But it should be singular here in the message I think :)